core/chain_makers.go

```diff
@@ -239,7 +243,83 @@ func GenerateChain(config *params.ChainConfig, parent *types.Block, engine conse
 	return blocks, receipts
 }
 
-func makeHeader(chain consensus.ChainReader, parent *types.Block, state *state.StateDB, engine consensus.Engine) *types.Header {
+type knot struct {
+	block *types.Block
+}
+
+func GenerateKnot(config *params.ChainConfig, parent *types.Block, engine consensus.Engine, db ethdb.Database, n int, gen func(int, *BlockGen)) ([]*types.Block, []types.Receipts) {
+	if config == nil {
+		config = params.TestChainConfig
+	}
+	blocks, receipts := make(types.Blocks, n), make([]types.Receipts, n)
+	chainreader := &fakeChainReader{config: config}
+
+	resultLoop := func(i int, k *knot, statedb *state.StateDB, results chan *types.Block, stop chan struct{}, wg *sync.WaitGroup) error {
+		for {
+			select {
+			case block := <-results:
+				context, err := engine.GetDifficultyOrder(block.Header())
+				if err != nil {
+					log.Println("Block mined has an invalid order")
+				}
+
+				if context == 0 {
+					log.Println("PRIME:", block.Header().Location, block.Header().Number, block.Header().Hash())
+				}
+
+				k.block = block
+
+				defer wg.Done()
+
+				return nil
+			}
+		}
+	}
+
+	genblock := func(i int, parent *types.Block, statedb *state.StateDB, location []byte, resultCh chan *types.Block, stopCh chan struct{}) {
+		b := &BlockGen{i: i, chain: blocks, parent: parent, statedb: statedb, config: config, engine: engine}
+		b.header = makeHeader(chainreader, parent, statedb, b.engine, location)
+
+		// Execute any user modifications to the block
+		if gen != nil {
+			gen(i, b)
+		}
+		if b.engine != nil {
+			// Finalize and seal the block
+			block := types.NewBlock(b.header, nil, nil, nil, trie.NewStackTrie(nil))
```

Comment:
After making the block, I think you need to add coinbase outputs & state transitions for each block. Unless we want to have an exception that says the genesis knot doesn't generate coinbases

Reply:
I was having problems with regenerating the state root so I left them as nil for now.